changes/2020-07-15_raw-rsa-keyring-public-private-key-validation/change.md

```diff
@@ -0,0 +1,92 @@
+
+This serves as a reference of all features that this change affects.
+
+| Feature                                                                                                                                                   |
+| --------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [Raw RSA Keyring](https://github.com/awslabs/aws-encryption-sdk-specification/blob/31b0534c4259aad365f048b73231545583389c67/framework/raw-rsa-keyring.md) |
+
+## Affected Specifications
+
+This serves as a reference of all specification documents that this change affects.
+
+| Specification                                                                                                                                             |
+| --------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| [Raw RSA Keyring](https://github.com/awslabs/aws-encryption-sdk-specification/blob/31b0534c4259aad365f048b73231545583389c67/framework/raw-rsa-keyring.md) |
+
+## Affected Implementations
+
+This serves as a reference for all implementations that this change affects.
+
+| Language   | Repository                                                                            |
+| ---------- | ------------------------------------------------------------------------------------- |
+| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
+| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |
+
+## Definitions
+
+- "match", "matching pair" describe an interoperable RSA public key, and RSA private key
+  (i.e. a ciphertext encrypted by the public key can be decrypted by the private key)
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+Add a requirement to the Raw RSA Keyring that if both the public key and the private key are defined,
+they MUST form a matching pair.
+
+Clarify language in OnEncrypt and OnDecrypt to require failure, instead of the vague "MUST NOT succeed".
+
+## Out of Scope
+
+- All keyrings aside from the Raw RSA Keyring
+- The manner in which the Raw RSA Keyring is build/constructed/instantiated
+- How to determine if a pair of keys match
+
+## Motivation
+
+A Raw RSA Keyring with mismatched keys is by construction unable to fulfill its decryption contract.
+As such, we should not allow such a keyring to be constructed.
+
+Language clarification around failure is needed,
+as previous language "MUST NOT succeed" could imply a non-failure response is valid
+(such as returning encryption/decryption materials unmodified).
```

Comment:
We can flip this language and say "We MUST only succeed if the keyring was initialized with matching pairs" or something similar and solve this problem without needing to check if the pairs match at initialization.

Reply:
From @seebees 's comment above, not every language has a way to do this easily. I'm incredibly worried this will be both burdensome to ESDK developers and be prone to bugs in complex implementations in order to satisfy this requirement.

Reply:
I don't think any runtime has a way to do this easily. If we don't want to require the check, we should have a conversation about whether we want one at all.